src/main/java/org/hibernate/infra/asciidoctor/extensions/customnumbering/CustomNumberingProcessor.java

```diff
@@ -83,33 +90,32 @@ private RubyObject toRubyObject(AbstractBlock block) {
 		}
 	}
 
-	private static class NumberingIndexes {
+	private SectionNumberingIndexes getSectionNumberingIndexes(int sectionNumber) {
+		return sectionNumberingIndexesMap.computeIfAbsent( sectionNumber, sn -> new SectionNumberingIndexes( sn ) );
+	}
 
-		private AtomicInteger example;
-		private AtomicInteger table;
-		private final String sectionNumber;
+	private static class SectionNumberingIndexes {
 
-		private NumberingIndexes(String sectionNumber) {
-			this.sectionNumber = sectionNumber;
-			example = new AtomicInteger( 1 );
-			table = new AtomicInteger( 1 );
-		}
+		private final int sectionNumber;
+		private AtomicInteger exampleIndex;
+		private AtomicInteger tableIndex;
```

Comment:
:) I kept them as well - at first these counters were not encapsulated and passed as method parameters for recursive methods. But now we can change it to simple ints.